internal/api/grpc/server/middleware/execution_interceptor.go

```diff
@@ -0,0 +1,190 @@
+package middleware
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+
+	"github.com/zitadel/logging"
+	"google.golang.org/grpc"
+
+	"github.com/zitadel/zitadel/internal/api/authz"
+	"github.com/zitadel/zitadel/internal/domain"
+	"github.com/zitadel/zitadel/internal/execution"
+	"github.com/zitadel/zitadel/internal/query"
+	exec_repo "github.com/zitadel/zitadel/internal/repository/execution"
+	"github.com/zitadel/zitadel/internal/telemetry/tracing"
+)
+
+func ExecutionHandler(queries *query.Queries) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		requestTargets, responseTargets := queryTargets(ctx, queries, info.FullMethod)
+
+		// call targets otherwise return req
+		handledReq, err := executeTargetsForRequest(ctx, requestTargets, info.FullMethod, req)
+		if err != nil {
+			return nil, err
+		}
+
+		response, err := handler(ctx, handledReq)
+		if err != nil {
+			return nil, err
+		}
+
+		return executeTargetsForResponse(ctx, responseTargets, info.FullMethod, handledReq, response)
+	}
+}
+
+func executeTargetsForRequest(ctx context.Context, targets []execution.Target, fullMethod string, req interface{}) (_ interface{}, err error) {
+	ctx, span := tracing.NewSpan(ctx)
+	defer span.EndWithError(err)
+
+	// if no targets are found, return without any calls
+	if len(targets) == 0 {
+		return req, nil
+	}
+
+	ctxData := authz.GetCtxData(ctx)
+	info := &ContextInfoRequest{
+		FullMethod: fullMethod,
+		InstanceID: authz.GetInstance(ctx).InstanceID(),
+		ProjectID:  ctxData.ProjectID,
+		OrgID:      ctxData.OrgID,
+		UserID:     ctxData.UserID,
+		Request:    req,
+	}
+
+	request, err := execution.CallTargets(ctx, targets, info)
+	// error gets only returned if InterruptOnError is set, or internal errors occur
+	if err != nil {
+		// if an error is returned still return also the original request
+		return req, err
+	}
+	return request, nil
```

Comment:
> if an error is returned still return also the original response

Why? The request is ignored by the caller if the error is not nil, right?

```suggestion
	return execution.CallTargets(ctx, targets, info)
```

Reply:
you're right, I added additional tests